backend/src/supervisors/supervisor.py

```diff
@@ -1,30 +1,86 @@
+from src.agents import DatastoreAgent, GoalAchievedAgent, UnresolvableTaskAgent, MathsAgent
+
+unresolvable_task_agent = UnresolvableTaskAgent()
+goal_achieved_agent = GoalAchievedAgent()
+list_of_agents = [ DatastoreAgent(), MathsAgent(), goal_achieved_agent, unresolvable_task_agent ]
+
+prompt_engine = PromptEngine()
+
+
+def still_solving_tasks(current_agent, iter_count):
+    agent_calling_unresolved = current_agent != "UnresolvableTaskAgent" and current_agent != "GoalAchievedAgent"
+    count_not_exceeded = iter_count < 5
+    return agent_calling_unresolved and count_not_exceeded
+
+
+def onto_next_task(best_next_step_json):
+    return best_next_step_json["current_or_next_task"].__contains__("next")
+
+
+def find_agent_from_name(current_agent_name):
+    return (agent for agent in list_of_agents if agent.name == current_agent_name)
 
 
 def solve_all_tasks(tasks_json):
-    first_task = str(tasks_json["tasks"][0])
-    logging.info("solving first task: " + first_task)
-
-    agent_for_first_task = pick_agent(first_task)
-    logging.info("agent picked: " + agent_for_first_task)
-
-    stub_final_answer = "Something went wrong. The agent picked was " + agent_for_first_task + " which does not exist"
-
-    # TODO: This is a temp setup. Adjust to call an agent's pick_method method
-    if agent_for_first_task == "database_agent":
-        # will call agent here
-        stub_final_answer = "I have used database_agent to retrieve personal information about you"
-    if agent_for_first_task == "financial_advisor_agent":
-        # will call agent here
-        stub_final_answer = "I have called the fiancial_advisor_agent to aid in the first task"
-    if agent_for_first_task == "web_search_agent":
-        # will call agent here
-        stub_final_answer = "I have searched the web with web_search_agent to find data for the first task"
-    if agent_for_first_task == "unresolvable_agent":
-        # will call agent here
-        stub_final_answer = "I was unable to find an agent to solve the specific task."
-
-    stub_final_answer = stub_final_answer + ' Task was "' + first_task + '"'
-
-    return stub_final_answer
+    current_agent_name = "initial_agent"
+    history = []
+    attempts_count = 0
+    task_step = 0
+    task_total_count = len(tasks_json["tasks"])
+    logging.info("Solving all (" + str(task_total_count) + ") tasks")
+
+    # Check where we should keep iterating to call more agents
+    while still_solving_tasks(current_agent_name, attempts_count):
+        logging.info(f"attempts_count: {attempts_count}, task_step: {task_step}")
+        logging.info(f"current_agent: {current_agent_name}, history: {history}")
+
+        # Assign tasks
+        current_task = str(tasks_json["tasks"][task_step])
+
+        # If on the last task don't iterate
+        if (task_step == task_total_count-1):
+            next_task = "There is no next task - the Next Task is the final task (sorry I know this is confusing)"
```

Comment:
Sorry Charlie, who is this message to? The LLM?

Reply:
It was a note I left in haste when frantically creating this bit of logic, hence the humor. The message is for the LLM to explicitly tell it to just solve the current task and not the next task.

I have refactored this to be clearer